plugins/util/portmap/portmap.go

```diff
@@ -0,0 +1,223 @@
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/utils"
+	"github.com/containernetworking/cni/pkg/utils/sysctl"
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/vishvananda/netlink"
+)
+
+// forwardPorts creates iptables rules to forward traffic from ports on
+// If enableSNAT is set, it also creates a masquerading chain that rewrites
+// connections to localhost:<port> so they also work.
+func forwardPorts(config *PortMapConf, containerIP net.IP, enableSNAT bool) error {
+	isV6 := (containerIP.To4() == nil)
+
+	var ipt *iptables.IPTables
+	var err error
+
+	if isV6 {
+		ipt, err = iptables.NewWithProtocol(iptables.ProtocolIPv6)
+	} else {
+		ipt, err = iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to open iptables:", err)
+	}
+
+	dnatChain := genDnatChain(config.Name, config.ContainerID)
+	dnatRules := dnatRules(config.RuntimeConfig.PortMaps, containerIP)
+	if err := dnatChain.setup(ipt, dnatRules); err != nil {
+		return fmt.Errorf("unable to setup DNAT", err)
+	}
+
+	if enableSNAT {
+		snatChain := genSnatChain(config.Name, config.ContainerID, isV6)
+		snatRules := snatRules(config.RuntimeConfig.PortMaps, containerIP)
+		if err := snatChain.setup(ipt, snatRules); err != nil {
+			return fmt.Errorf("unable to setup SNAT", err)
+		}
+		if !isV6 {
+			hostIfName := getRoutableHostIF(config.PrevResult)
+			if hostIfName != "" {
+				if err := enableLocalnetRouting(hostIfName); err != nil {
+					return fmt.Errorf("unable to enable route_localnet: %v", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func genDnatChain(netName, containerID string) chain {
+	name := utils.FormatChainName("DN-", netName, containerID)
+	comment := fmt.Sprintf(`dnat name: "%s" id: "%s"`, netName, containerID)
+
+	return chain{
+		name: name,
+		entryRule: []string{
+			"-m", "addrtype",
+			"--dst-type", "LOCAL",
+			"-m", "comment",
+			"--comment", comment,
+			"-j", name},
+		entryChains: []string{"PREROUTING", "OUTPUT"},
+	}
+}
+
+// dnatRules generates the destination NAT rules, one per port, to direct
+// traffic from hostip:hostport to podip:podport
+func dnatRules(entries []PortMapEntry, containerIP net.IP) [][]string {
+	out := make([][]string, 0, len(entries))
+	for _, entry := range entries {
+		out = append(out, []string{
+			"-p", entry.Protocol,
+			"--dport", strconv.Itoa(entry.HostPort),
+			"-j", "DNAT", "--to-destination",
+			fmtIpPort(containerIP, entry.ContainerPort),
+		})
+	}
+	return out
+}
+func genSnatChain(netName, containerID string, isV6 bool) chain {
+	name := utils.FormatChainName("SN-", netName, containerID)
+	comment := fmt.Sprintf(`snat name: "%s" id: "%s"`, netName, containerID)
+
+	if isV6 {
+		return chain{
+			name: name,
+			entryRule: []string{
+				"-s", "::1",
+				"!", "-d", "::1",
+				"-m", "comment",
+				"--comment", comment,
+				"-j", name},
+			entryChains: []string{"POSTROUTING"},
```

Comment:
We can't assume that we're in a k8s world with the MARK-MASQ setup. However, it can be provided as an option.